dulwich/repo.py

```diff
@@ -1080,7 +1090,21 @@ def stage(self, fs_paths):
             else:
                 if not stat.S_ISDIR(st.st_mode):
                     blob = blob_from_path_and_stat(full_path, st)
-                    blob = blob_normalizer.checkin_normalize(blob, fs_path)
+
+                    if not head_tree:
+                        new_file = True
+                    else:
+                        try:
+                            head_tree.lookup_path(lookup_obj, tree_path)
```

Comment:
Are you sure git actually looks in HEAD? "git add" is supposed to be an index-only operation.

Reply:
I tried adding a new file with CRLF line-ending to the staging area and it was correctly converted into LF (with autocrlf input). Then I tried modifying it by adding new CRLF and when adding the new content, it also got converted.
When introducing new CRLF lines in an already tracked file, CGit seems to not run the line-ending conversion so I suspect CGit has a way to detect when it's a new file or not.
My guess is it either check the store on the current branch (I tried adding a file already tracked on another branch and it got converted) or it saves some extra information in the index (saying that file X is new) and discard this information after a commit. Do you have ideas how we could find how CGit is doing it?

Reply:
After more test, maybe CGit is comparing the line-endings between the working directory and the index and apply the conversion if the file in index does not contains CRLF and the file in working directory does, I will try to validate this hypothesis